README
A virtual machine for the IPS language on Unix/Linux systems
============================================================
Copyright (C) 2003, Pieter-Tjerk de Boer -- pa3fwm@amsat.org


Introduction
============

IPS (Interpreter for Process Structures) is a somewhat Forth-like language
developed in the 1970s by Dr Karl Meinzer, mainly aimed at small computer
systems that control equipment.
It has been used for the on-board computer on several AMSAT (Amateur Radio)
spacecraft (AO-10, AO-13 and AO-40), and is also planned to be used on future
AMSAT spacecraft.
Versions of IPS have also been developed for use on desk-top computers,
and the present package contains such a version that should run on just
about any Linux or other Unix-like system.


Getting started
===============

First, compile the system, typically by simply typing 'make' at your unix
command prompt. You may need to slightly modify the contents of Makefile
if e.g. libraries cannot be found or your C compiler needs different options
than gcc.
If your computer is NOT a standard IBM/Intel-compatible PC, you should check
whether its CPU is big-endian or little-endian. If it is big-endian, you
must add  -DBIGENDIAN  at the end of the CFLAGS line in the Makefile.
If in doubt, it is also safe to do so since that will actually work correctly
on both types of machine; however, on little-endian machines it is more
efficient to leave it out (as is the default, since the large majority of
users probably have a PC).

Next, start the resulting binary by typing  ./ips .
If you want to run it in a different directory, you also need to copy the file
IPS-Mu.bin to that directory, or use the -i option to ips; see ips -h for more
detail.

You now have the IPS console on your screen, and can start typing IPS commands.
For example, perform a simple addition by typing
  23 45 +
(followed by <enter>).
You can define new IPS commands, for example a command to calculate factorials:
  : factorial DUP 1 <> JA? DUP 1 - factorial * DANN ;
Then
  5 factorial
will calculate 5*4*3*2*1=120
(Note however that for numbers larger than 7, the answer is not correct
since IPS works with signed 16-bit numbers.)

Two example IPS programs have been included: bigclock.ips and gs2003.ips.
The former displays a big clock, the latter is a satellite tracking program
loaded with the keppler data for AO-40.
You can load one of these by typing a command like
  " filename.ips " READ

The IPS system is closed by simply pressing the interrupt key, which is
typically Ctrl-C .

If you want to do more with IPS, you'll probably want to study the IPS book.


Background and implementation notes
===================================

The IPS system for Linux/Unix consists of three parts:
- The ``memory image'' of the system; this is the actual IPS kernel; it is
  contained in the file IPS-Mu.bin.
- A virtual machine that can interpret the above-mentioned memory image;
  this part is written in C, in the file emu.c .
- The user interface, which handles file I/O, reads the keyboard, and displays
  the IPS system's 64x16 character screen on a standard terminal (console or
  xterm), and initially sets the IPS system clock.
  This part is also written in C, in the file ips.c .
Note that in the "real" spacecraft implementations, the latter two parts are
not present: the memory image there is coded in the spacecraft CPU's native
machine code so no interpreter in the form of a virtual machine is needed,
and the keyboard and screen are replaced by telemetry up- and downlinks,
which are partially implemented in hardware and partially in software (in the
CPU's native machine code, inside the memory image).

Now you may wonder where the memory image IPS-Mu.bin comes from.
Well, its source code is included in the package: that is the file IPS-Mu.SRC .
As you can see, this is itself written in a rather IPS-like language, which
is called IPS-X.
In order to generate IPS-Mu.bin from IPS-Mu.SRC, the same virtual machine can
be used, but it needs to be loaded with a different memory image. This alternate
memory image is in the file IPS-Xu.bin, and will be loaded by invoking the
system as  ips -x . Thus, IPS-Mu.bin can be generated from IPS-Mu.SRC as follows:
- start  ips -x
- type  ~ IPS-Mu.SRC ~ read (note the use of tilde (~) instead of quote ("))
  This step will (over)write IPS-Mu.bin .
- press Ctrl-C to stop the system.

Next obvious question of course is, where does the cross-compile image IPS-Xu.bin
come from? Well, it was simply copied from Paul Wilmott's IPSDOS package (where
it was called IPS-XP.BIN).
That's the advantage of using virtual machines rather than writing the memory
image in native machine language: the memory images can be compatible among
different systems; in particular, IPS-M for Acorn, for DOS, for Windows and
for Linux/unix are compatible. 
Of course, IPS-XP.BIN does have a history of its own, but for that you should
look at the IPSDOS package.

As noted above, the virtual machines for several platforms are compatible.
As a consequence, not only the cross-compiler image could be shared among
platforms, but also the regular IPS image. However, in IPS-Mu.SRC (and thus
also in IPS-Mu.bin), some minor changes have been made, compared to their
counterparts for e.g. Acorn or DOS. These changes are:
- IPS-Mu does not have some IPS instructions that don't make sense on the
  Linux/Unix implementation, such as the DEFCHAR instruction from Acorn and
  the hardware I/O instructions from the DOS version.
- IPS-Mu contains a minor change which helps the Linux/Unix virtual machine,
  to detect whether the IPS system is just running an idle loop, and if so,
  give up some CPU time to the host operating system.
  In native implementations of IPS on board of e.g. spacecraft, IPS is the
  only process running on the CPU, so it can just use all CPU cycles. However,
  when IPS is running as one of many tasks on a Linux/Unix system, it is
  annoying when IPS takes lots of CPU cycles while idling; hence this change.

More information
================
For more information, you may want to look at the following:
- the book "IPS - High Level Programming of Small Systems" by Karl Meinzer,
  1978, second edition 1997, ISBN 0-9530507-0-X.
- http://www.jrmiller.demon.co.uk/products/ipsbk.html
- http://www.amsat.org/amsat/sats/ao40/ips.html



ChangeLog
150630AP
	Makefile.win32
	original ips.c renamed to ips_unix.c, PDCurses fails to translate
	KEY_BACKSPACE/Ctrl-C hence the Mingw port
	ips_win32.c created using SF conio21 library, ipsw.exe


curterm.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "curterm.h"

static int ctrl_c_event;

#ifdef __MINGW32__
#define WIN32_LEAN_AND_MEAN 1
#include <windows.h>
#include <conio.h>

static HANDLE hStdin;
static HANDLE hStdout;

void gotoxy(int x, int y)
{
	COORD xy;

	xy.X = x;
	xy.Y = y;
	SetConsoleCursorPosition(hStdout, xy);
}

void clrscr(void)
{
	CONSOLE_SCREEN_BUFFER_INFO scr;
	COORD ul = {0, 0};
	DWORD n, size;

	GetConsoleScreenBufferInfo(hStdout, &scr);
	size = scr.dwSize.X * scr.dwSize.Y;
	FillConsoleOutputCharacter(hStdout, ' ', size, ul, &n);
	FillConsoleOutputAttribute(hStdout,   0, size, ul, &n);
	SetConsoleCursorPosition(hStdout, ul);
}

BOOL CtrlC(DWORD ct)
{
	if (CTRL_C_EVENT == ct) {
		ctrl_c_event = 1;
		return TRUE;
	}
	return FALSE;
}

void initterm()
{
	static int _init = 0;

	if (!_init) {
		hStdin  = GetStdHandle(STD_INPUT_HANDLE);
		hStdout = GetStdHandle(STD_OUTPUT_HANDLE);
		_init = 1;
	}
}

void cbreak(int enable)
{
	initterm();

	ctrl_c_event = 0;
	if (!SetConsoleCtrlHandler((PHANDLER_ROUTINE)CtrlC, enable ? FALSE : TRUE)) 
	{
		printf("failed to install Ctrl-C handler\n");
		exit(1);
	}

}

void prepterm(int dir)
{
	static DWORD told, tnew;

	initterm();

	if (dir) {
		GetConsoleMode(hStdin, &told);
		tnew  = told;
		tnew &= ~ENABLE_PROCESSED_INPUT;
		SetConsoleMode(hStdin, tnew);
	}
	else
		SetConsoleMode(hStdin, told);

	return;
}

int has_key()
{
	return kbhit();
}

int getkey(void)
{
	int ch;

	ch = getch();
	if (0 == ch)
		ch = 0xF0;
	if (0xF0 == ch || 0xE0 == ch)
		ch = (ch << 8) + getch();

	return ch;
}

#else

#include <termios.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/time.h>

void prepterm(int dir)
{
	static struct termios told, tnew;

	if (dir) {
		tcgetattr(STDIN_FILENO, &told);
		tnew = told;
		tnew.c_lflag &= ~(ICANON | ECHO );
		tcsetattr(STDIN_FILENO, TCSANOW | TCSAFLUSH, &tnew);
	}
	else
		tcsetattr(STDIN_FILENO, TCSANOW | TCSAFLUSH, &told);
}

int has_key(void)
{
	struct timeval tv;
	fd_set rd;

	tv.tv_sec  = 0;
	tv.tv_usec = 0;

	FD_ZERO(&rd);
	FD_SET(STDIN_FILENO, &rd);

	select(STDIN_FILENO + 1, &rd, NULL, NULL, &tv);
	return FD_ISSET(STDIN_FILENO, &rd);
}

#define ESC '\033'

int getkey(void)
{
	int ch;

	ch = getchar();
	/* NOTE: cannot read ESC key */
	if (ESC == ch) {
		ch = (u_char)getchar();
		ch = (ch << 8) + (u_char)getchar();
	}

	return ch;
}

void gotoxy(int x, int y)
{
	char buf[16];

	sprintf(buf, "%c[%d;%dH", ESC, y, x);
	write(1, buf, strlen(buf));
}

void clrscr(void)
{
	char buf[16];

	sprintf(buf, "%c[2J", ESC);
	write(1, buf, strlen(buf));
	gotoxy(0, 0);
}

void ctrlc_handler(int sig)
{
	ctrl_c_event = 1;
}

void cbreak(int enable)
{
	if (enable)
		signal(SIGINT, SIG_DFL);
	else
		if (SIG_ERR == signal(SIGINT, ctrlc_handler)) {
			printf("failed to install Ctrl-C handler\n");
			exit(1);
		}
}

#endif

#ifdef MAIN
int main(int argc, char *argv[])
{
	int c;

	c = 0;
	prepterm(1);
	while (c != 'q') {
		if (has_key()) {
			printf("%04X ", (c = getkey()));
			if (27 == c) {
				printf("%d ", (c = getkey()));
				printf("%d ", (c = getkey()));
			}
			printf("\n");
		}
	}
	prepterm(0);
	return 0;
}
#endif

void kbflush(void)
{
	while (has_key())
		getkey();
}

int has_ctrlc(void)
{
	if (ctrl_c_event) {
		ctrl_c_event = 0;
		return 1;
	}
	return 0;
}


ips_win32.c

/*  IPS - virtual machine for the IPS (Interpreter for Process Structures)
 *  language / computing environment used onboard of AMSAT spacecraft
 *
 *  Copyright (C) 2003, Pieter-Tjerk de Boer -- pa3fwm@amsat.org
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of version 2 of the GNU General Public License as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *
 *  This module contains the unix- or ncurses-specific parts of the
 *  program; i.e., the processing of command-line options, the user
 *  interface and the timing stuff.
 *  It also contains the few lines of code that load the actual IPS
 *  binary image into the emulator's memory mem[] .
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/timeb.h>
#include <time.h>
#include <unistd.h>

#include <conio2.h>
#define WIN32_LEAN_AND_MEAN	1
#include <windows.h>

#include "ips.h"


/**************** timing-related code **********************/

void do_sleep(void)   /* sleep for a little while, i.e., give up CPU time to other processes */
{
	_sleep(10);
}


int test_20ms(void)    /* test whether another 20 milliseconds have passed by */
{
	struct _timeb tb;
   static long thr_msec=0;
	_ftime(&tb);
   if (tb.millitm<thr_msec || tb.millitm>=thr_msec+20) {
      thr_msec+=20;
      if (thr_msec==1000) thr_msec=0;
      return 1;
   }
   return 0;
}


void init_uhr(void)   /* initialize the UHR with the present time */
{
   time_t t;
   struct tm *tm;

   time(&t);
   tm=gmtime(&t);
   pokeB(UHR,0);
   pokeB(UHR+1,tm->tm_sec);
   pokeB(UHR+2,tm->tm_min);
   pokeB(UHR+3,tm->tm_hour);

   tm->tm_mon+=2;
   if (tm->tm_mon<4) { tm->tm_year--; tm->tm_mon+=12; }
   poke(UHR+4, (tm->tm_mon*306)/10 + tm->tm_year*365 + tm->tm_year/4 + tm->tm_mday - 28553 );
}



/**************** code for handling screen and keyboard **********************/

unsigned short norm_attr;
unsigned short rev_attr;
static HANDLE hStdin;
static DWORD old_mode;

void do_redraw(void)   /* refresh the screen */
{
   int i,j;
	struct char_info scr[16*64];
	unsigned short attr;

	_conio_gettext(2,2,64+1,16+1,scr);
   for (j=0;j<16;j++) {
      for (i=0;i<64;i++) {
         int c=mem[j*64+i];
         int inv=(c&0x80);
         if (inv) attr = rev_attr;
         else attr = norm_attr;
         c&=0x7f;
         if (c<0x20) c|=0x80;
			scr[j*64+i].letter = c;
			scr[j*64+i].attr   = attr;
      }
   }
	puttext(2,2,64+1,16+1,scr);
}

void init_console(void)
{
	int i;
	DWORD new_mode;
	struct text_info ti;

	hStdin  = GetStdHandle(STD_INPUT_HANDLE);

	GetConsoleMode(hStdin, &old_mode);
	new_mode = old_mode & ~ENABLE_PROCESSED_INPUT;
	SetConsoleMode(hStdin, new_mode);

	inittextinfo();
	gettextinfo(&ti);
	norm_attr = LIGHTGRAY + (     BLUE << 4);
	rev_attr  =      BLUE + (LIGHTGRAY << 4);

	textcolor(LIGHTGRAY);
	textbackground(BLUE);
	clrscr();
	cputsxy(1,1,
"+----------------------------------------------------------------+");
	for (i = 0; i < 16; i++)
		cputsxy(1,2+i,
"|                                                                |");
	cputsxy(1,18,
"+----------------------------------------------------------------+");
	cputsxy(18,1,"[Meinzer M-9097 IPS Computer]");
}

void end_console(void)
{
	normvideo();
	clrscr();
	SetConsoleMode(hStdin,old_mode);
	exit(0);
}

#define KEY_LEFT				0xE04B
#define KEY_RIGHT				0xE04D
#define KEY_UP					0xE048
#define KEY_DOWN				0xE050
#define KEY_BACKSPACE		0x0008
#define KEY_DC					0xE053
#define KEY_IC					0xE052	
#define CTRL_C					0x0003
#define ERR						0

int get_ch(void)
{
	int c = 0;

	if (kbhit()) {
		c = getch();
		if (0 == c || 0xE0 == c) {
			if (!c) c = 0xF0;
			c = (c << 8) + getch();
		}
	}
	return c;
}

int handle_keyboard(void)
{
   int c=0;
   static int insertmode=1;
   int ret=1;

   mem[input_ptr]&=0x7f;

	c = get_ch();

	if (CTRL_C == c)
		end_console();

   switch (c) {
      case ERR: 
         ret=0;
         break;
      case '\n':
      case '\r':
         return 2;
      case KEY_LEFT:
         input_ptr--;
         if (input_ptr<0) input_ptr=0;
         break;
      case KEY_UP:
         input_ptr-=64;
         if (input_ptr<0) input_ptr=0;
         break;
      case KEY_RIGHT:
         input_ptr++;
         if (input_ptr>TVE) input_ptr=TVE;
         break;
      case KEY_DOWN:
         input_ptr+=64;
         if (input_ptr>TVE) input_ptr=TVE;
         break;
      case KEY_BACKSPACE:
         input_ptr--;
         if (input_ptr<0) input_ptr=0;
      case KEY_DC:
         if (insertmode) {
            memmove(mem+input_ptr,mem+input_ptr+1,TVE-input_ptr);
            mem[TVE]=' ';
         } else {
            mem[input_ptr]=' ';
         }
         break;
      case KEY_IC:
         insertmode=!insertmode;
         break;
      default:
         if (insertmode) memmove(mem+input_ptr+1,mem+input_ptr,TVE-input_ptr);
         pokeB(input_ptr++,c); 
         if (input_ptr>TVE) input_ptr=TVE;
   }

   if (!peekB(READYFLAG)) mem[input_ptr]|=0x80;

   return ret;
}

/**************** startup code **********************/

int main(int argc,char **argv)
{
   FILE *f;
   char *cmd=NULL;
   char *image="IPS-Mu.bin";

   /* parse command line */
   while (1) {
      int i;
      i=getopt(argc,argv,"c:f:i:xh");
      if (i<0) break;
      switch (i) {
         case 'c':
            if (!optarg) goto error;
            cmd=optarg;
            break;
         case 'f':
            if (!optarg) goto error;
            cmd=malloc(strlen(optarg)+16);
            if (!cmd) goto error;
            sprintf(cmd,"\" %s \" READ",optarg);
            break;
         case 'i':
            if (!optarg) goto error;
            image=optarg;
            break;
         case 'x':
            image="IPS-Xu.bin";
            break;
         default:
error:
            printf("\n"
"IPS for unix/linux\n"
"------------------\n"
"Options are:\n"
"-c <commands>   - initial IPS commands\n"
"-f <filename>   - load IPS source file, equivalent to -c '\" <filename> \" READ'\n"
"-i <filename>   - load image from file, default is IPS-Mu.bin\n"
"-x              - be cross-compiler, equivalent to -i IPS-Xu.bin\n"
"");
            return 1;
      }
   }

   /* load the image */
   f=fopen(image,"r");
   if (!f) {
      fprintf(stderr,"Can't open image file \"%s\".\n",image);
      return 1;
   }
   fread(mem,1,65536,f);
   fclose(f);

   /* if commands were given on the command line, copy them into the screen memory as input */
   if (cmd) {
      if (strlen(cmd)>64*8) {
         fprintf(stderr,"Command too long (max. 512 bytes)\n");
         return 1;
      }
      memcpy(mem+64*8,cmd,strlen(cmd));
      poke(a_PE,0x200+strlen(cmd));
      mem[READYFLAG]=1;
   }

   /* initialize the ncurses stuff */
   init_console();

   /* initialize the UHR */
   init_uhr();

   /* finally, run the emulator */
   emulator();

   return 0;
}

